### DISCUSSION TITLE
Write more than three fields in PlotfileToTurb

### DISCUSSION THREAD

**Discussion by indra098124** [AMReX-Codes/amrex#3775] [Q&A]
Currently PlotfileToTurb.cpp writes three fields, I am looking to extend it to write more fields. I have difficulty in understanding the following code snippet. How do we decide the shift in vfab and what the index in vfab correspond to? 

`static
void
Extend (FArrayBox& xfab,
        FArrayBox& vfab,
        const Box& domain)
{
  Box tbx = vfab.box();

  tbx.setBig(0, domain.bigEnd(0) + 3);

  const int ygrow = BL_SPACEDIM==3 ? 3 : 1;

  tbx.setBig(1, domain.bigEnd(1) + ygrow);

  xfab.resize(tbx,1);

  xfab.copy(vfab);
  vfab.shift(0, domain.length(0));
  xfab.copy(vfab);
  vfab.shift(1, domain.length(1));
  xfab.copy(vfab);
  vfab.shift(0, -domain.length(0));
  xfab.copy(vfab);
}`

My understanding is that the codes reads field names[d] from the pltfile and copies that to xfab. I can add more fields in names[d] and adjust the loop counter to loop over the new fields. But looks like that is not sufficient. Could you please suggest what other changes will be needed to accomplish this? I also note that the code fails when using more than one mpi rank? Is this meant to run in serial only? 

Many thanks!

**Comment by WeiqunZhang** [comment]
The code shown above extends the original data by assuming the domain is periodic in x and y directions. It does one component at a time with a loop `for (int d = 0; d < BL_SPACEDIM; ++d)`. I suppose you can change the for loop if you have more than BL_SPACEDIM components. The PlotfileToTurb writes data specifically for IAMR's turbulence inflow. Is that what you want?

**Comment by indra098124** [comment]
Thanks @WeiqunZhang for looking into this. I want to map the scalar along with the velocities at the inlet. I modified 
std::string         names[5]    = { "x_velocity", "y_velocity", "z_velocity",  "temperature" };
and modified the loop for (int d = 0; d < BL_SPACEDIM+1; ++d)

I have difficultly in understanding what  Extend(xfab, TMP[0], domain) does and whether this should be modified as we increase the number of variables to map.

this is the code for extend 
`
static
void
Extend (FArrayBox& xfab,
        FArrayBox& vfab,
        const Box& domain)
{
  Box tbx = vfab.box();

  tbx.setBig(0, domain.bigEnd(0) + 3);

  const int ygrow = BL_SPACEDIM==3 ? 3 : 1;

  tbx.setBig(1, domain.bigEnd(1) + ygrow);

  xfab.resize(tbx,1);

  xfab.copy(vfab);
  vfab.shift(0, domain.length(0));
  xfab.copy(vfab);
  vfab.shift(1, domain.length(1));
  xfab.copy(vfab);
  vfab.shift(0, -domain.length(0));
  xfab.copy(vfab);
}`

I am not sure how do we decide the arguments of shift and what is this shift doing?

Additionally, I was under the assumption that this utilitiy assumes the flow to be periodic in z. Do we also need the flow to be periodic in x and y?

**Reply by WeiqunZhang** [reply]
As I said, the function extends the original data by assuming the data are periodic. Suppose the original data are on from [0,nx-1] in x-direction and [0,ny-1] in y-direction. The code creates a new fab that is from [0, nx-1+3] in x-direction and [0,ny-1+3] in y-direction and fill the new cells outside the original domain at (i,j) with the original data at (i-nx,j), or (i,j-ny), or (i-nx, j-ny), whichever is in the original domain. Does it make sense?

As for periodic conditions, I think the assumption is the turbulence data are periodic, but not really the simulation that use the turbulence data. @drummerdoc can probably tell you more.

**Reply by drummerdoc** [reply]
The Extend function pads the data parallel to the inflow face around the edges to allow the parabolic interpolant to fill data up to the physical bounds of the rectangular patch of data (indicated by the fields in the HDR header file) without needing to do the offset calc on the fly to get the data from the other side.  It assumes that the input data was computed periodically so that the padded data can be sourced from the opposite side of the domain - then the data written has everything necessary to compute the interpolant locally.  However, if the location of the requested interpolation lives physically beyond the specified domain of the turbulence file (in the directions parallel to the inlet face), no filling is done there -- admittedly, this is perhaps inconsistent.  Normal to the face, the turbulence data is assumed to be periodic, and a z-location in the data is computed using a (constant) convection velocity, where the coordinate is thus effectively wrapped in the z direction - that is, the turbulence is cycled through periodically at the speed specified by the input turbulence convection speed.

These things are a bit tricky to write down in text and are very arbitrary for a specific usage.  Let me know if you need more info.  But yes, you can add components to the interpolation as Weiqun suggests.  You just need to ensure that you've allocated sufficient memory in the temp data structure and in the planes of data stored by the interpolator logic.

**Reply by drummerdoc** [reply]
Also, the odd indexing you see in Extend is to read in data that is assumed to be cell-centered, yet treat it in the interpolator function as node centered.  Hope that helps.

**Comment by indra098124** [comment]
Thank you @drummerdoc and @WeiqunZhang for explaining this.
